File: packages/core/src/factories/TypeFactory.ts
import ts from "typescript";

export namespace TypeFactory {
  export const isFunction = (type: ts.Type): boolean =>
    getFunction(type) !== null;

  export const getFunction = (type: ts.Type) => {
    const node = type.symbol?.declarations?.[0];
    if (node === undefined) return null;

    return ts.isFunctionLike(node)
      ? node
      : ts.isPropertyAssignment(node) || ts.isPropertyDeclaration(node)
        ? ts.isFunctionLike(node.initializer)
          ? node.initializer
          : null
        : null;
  };

  export const getReturnTypeOfClassMethod = (props: {
    checker: ts.TypeChecker;
    class: ts.Type;
    function: string;
  }): ts.Type | null => {
    // FIND TO-JSON METHOD
    const symbol: ts.Symbol | undefined = props.class.getProperty(
      props.function,
    );
    if (!symbol) return null;
    else if (!symbol.valueDeclaration) return null;

    // GET FUNCTION DECLARATION
    const functor: ts.Type = props.checker.getTypeOfSymbolAtLocation(
      symbol,
      symbol.valueDeclaration,
    );

    // RETURNS THE RETURN-TYPE
    const signature: ts.Signature | undefined =
      props.checker.getSignaturesOfType(functor, ts.SignatureKind.Call)[0];
    return signature ? signature.getReturnType() : null;
  };

  export const getFullName = (props: {
    checker: ts.TypeChecker;
    type: ts.Type;
    symbol?: ts.Symbol;
    aliasTypeArguments?: boolean; // default: true
  }): string => {
    // PRIMITIVE
    const symbol =
      props.symbol ?? props.type.aliasSymbol ?? props.type.getSymbol();
    if (symbol === undefined) return props.checker.typeToString(props.type);

    // UNION OR INTERSECT
    if (
      props.type.aliasSymbol === undefined &&
      props.type.isUnionOrIntersection()
    ) {
      const joiner: string = props.type.isIntersection() ? " & " : " | ";
      return props.type.types
        .map((child) =>
          getFullName({
            checker: props.checker,
            type: child,
          }),
        )
        .join(joiner);
    }

    //----
    // SPECIALIZATION
    //----
    const name: string = get_name(symbol);

    // CHECK GENERIC
    const generic: readonly ts.Type[] =
      props.type.aliasSymbol && props.aliasTypeArguments !== false
        ? (props.type.aliasTypeArguments ?? [])
        : props.checker.getTypeArguments(props.type as ts.TypeReference);
    return generic.length
      ? name === "Promise"
        ? getFullName({
            checker: props.checker,
            type: generic[0]!,
          })
        : `${name}<${generic
            .map((child) =>
              getFullName({
                checker: props.checker,
                type: child,
              }),
            )
            .join(", ")}>`
      : name;
  };

  const explore_name = (props: { node: ts.Node; name: string }): string =>
    ts.isModuleBlock(props.node)
      ? explore_name({
          node: props.node.parent.parent,
          name: `${props.node.parent.name.getFullText().trim()}.${props.name}`,
        })
      : props.name;

  const get_name = (symbol: ts.Symbol): string => {
    const parent = symbol.getDeclarations()?.[0]?.parent;
    return parent
      ? explore_name({
          node: parent,
          name: symbol.escapedName.toString(),
        })
      : "__type";
  };

  export const keyword = (
    type:
      | "void"
      | "any"
      | "unknown"
      | "boolean"
      | "number"
      | "bigint"
      | "string",
  ) => {
    return ts.factory.createKeywordTypeNode(
      type === "void"
        ? ts.SyntaxKind.VoidKeyword
        : type === "any"
          ? ts.SyntaxKind.AnyKeyword
          : type === "unknown"
            ? ts.SyntaxKind.UnknownKeyword
            : type === "boolean"
              ? ts.SyntaxKind.BooleanKeyword
              : type === "number"
                ? ts.SyntaxKind.NumberKeyword
                : type === "bigint"
                  ? ts.SyntaxKind.BigIntKeyword
                  : ts.SyntaxKind.StringKeyword,
    );
  };
}

File: packages/core/src/factories/TypeFactory.go
package factories

import "strings"

type typeFactory struct{}

var TypeFactory typeFactory

func (typeFactory) IsFunction(typ TypeLike) bool {
	return TypeFactory.GetFunction(typ) != nil
}

func (typeFactory) GetFunction(typ TypeLike) FunctionLike {
	if typ == nil {
		return nil
	}
	return typ.Function()
}

func (typeFactory) GetReturnTypeOfClassMethod(props ReturnTypeOfClassMethodProps) TypeLike {
	if props.Class == nil || props.Checker == nil {
		return nil
	}
	symbol := props.Class.Property(props.Function)
	if symbol == nil {
		return nil
	}
	return props.Checker.ReturnTypeOf(symbol)
}

func (typeFactory) GetFullName(props FullNameProps) string {
	if props.Type == nil {
		return "__type"
	}
	symbol := props.Symbol
	if symbol == nil {
		symbol = props.Type.AliasSymbol()
	}
	if symbol == nil {
		symbol = props.Type.Symbol()
	}
	if symbol == nil {
		return props.Type.String()
	}
	if props.Type.IsUnionOrIntersection() && props.Type.AliasSymbol() == nil {
		joiner := " | "
		if props.Type.IsIntersection() {
			joiner = " & "
		}
		names := make([]string, 0, len(props.Type.Types()))
		for _, child := range props.Type.Types() {
			names = append(names, TypeFactory.GetFullName(FullNameProps{Checker: props.Checker, Type: child}))
		}
		return strings.Join(names, joiner)
	}
	name := symbol.Name()
	generic := props.Type.TypeArguments()
	if len(generic) == 0 {
		return name
	}
	if name == "Promise" {
		return TypeFactory.GetFullName(FullNameProps{Checker: props.Checker, Type: generic[0]})
	}
	names := make([]string, 0, len(generic))
	for _, child := range generic {
		names = append(names, TypeFactory.GetFullName(FullNameProps{Checker: props.Checker, Type: child}))
	}
	return name + "<" + strings.Join(names, ", ") + ">"
}

func (typeFactory) Keyword(kind string) TypeNode {
	switch kind {
	case "void", "any", "unknown", "boolean", "number", "bigint", "string":
		return Node("keywordType", Field("type", kind))
	default:
		return Node("keywordType", Field("type", "any"))
	}
}
